README.md
bolo - A Versatile Monitoring Toolkit
=====================================

bolo is a toolkit for building distributed, scalable monitoring
systems.  It provides a set of small, versatile components that
you can plug together in new and exciting ways, to gather and
aggregate metrics, mine data and track system states.

Getting Started
---------------

To compile the software, use the standard incantation:

    $ ./bootstrap
    $ ./configure
    $ make
    $ make check           # optional, but recommended
    $ sudo make install

This will compile and install all of the bolo components in
standard systems places.  bolo requires libzmq3, pthreads, librrd,
libpcre3 and [libvigor][libvigor] to run.  It uses [ctap][ctap] to
run automated tests (which are, again, highly recommended).

A Minimal System
----------------

The core of every bolo deployment is the `bolo` daemon itself,
which requires a configuration file, usually **/etc/bolo.conf**.

    # /etc/bolo.conf
    #
    listener   tcp://*:2999
    controller tcp://127.0.0.1:2998
    broadcast  tcp://*:2997

    log info daemon
    savefile /var/lib/bolo/save.db
    keysfile /var/lib/bolo/keys.db

    dumpfiles /var/tmp/bolo.%s

    type :generic {
      freshness 60
      critical "no result from monitored thing"
    }
    state :generic m/./

    window @minutely 60
    use @minutely
    sample  m/./
    counter m/./
    rate    m/./

This is a **catch-all** configuration, that excepts all submitted
metrics to be collected at at least a 1-minute resolution, and
aggregate at the per-minute mark.  Check **bolo.conf(5)** for
details and more advanced usage.

You'll have to create the bolo user (or adjust how you launch the
bolo daemon):

    $ sudo useradd -rUd /var/lib/bolo -s /sbin/nologin bolo
    $ sudo mkdir /var/lib/bolo
    $ sudo chown -R bolo:bolo /var/lib/bolo

With this in place, start up the bolo daemon:

    $ sudo bolo

By default, bolo starts up daemonized, and reads from
/etc/bolo.conf.

Submitting Data
---------------

The **send_bolo** utility can be used to submit metrics and state
data to bolo, via the _listener_ port (TCP/2999 in our case):

    $ send_bolo -e tcp://localhost:2999 -t sample life 42
    $ send_bolo -e tcp://localhost:2999 -t sample universe 42
    $ send_bolo -e tcp://localhost:2999 -t sample everything 42

At this point, you will probably want to go get
[bolo-collectors][collectors] installed to give you some real
metric collectors, for real things like memory usage, swap rates,
CPU time allocation, etc.  The output from these collectors can be
piped directly into **send_bolo**:

    $ /path/to/collector | send_bolo -e tcp://localhost:2999

With collectors installed, you can also run **dbolo**, the
distributed bolo agent.  Create the _/etc/dbolo.conf_
configuration file:

    # /etc/dbolo.conf
    @10s /usr/lib/bolo/collectors/linux
    @1h  /usr/lib/bolo/collectors/hostinfo

And then launch dbolo as a standalone daemon:

    $ sudo dbolo

Now you've got a small agent process running and submitting
Linux-y performance data every 10s, and host metadata every hour.

Using the Data
--------------

As the bolo daemon receives metrics and state data from monitored
hosts, it aggregates the data and regularly broadcasts these
summaries out to connected subscribers.  One such subscriber is
**bolo2rrd**, which listens for samples, counters and rate data
and creates / updates RRD files on disk.

    $ sudo mkdir /srv/rrd
    $ sudo bolo2rrd -r /srv/rrd -e tcp://localhost:2997

bolo2rrd daemonizes into the background.

The **stat_bolo** utility also has a _listen_ mode wherein it
connects to the broadcast port and prints out the data as it is
received.  This can be useful for subscriber prototyping in
languages like Perl or Ruby, without having to deal directly with
0MQ and the bolo message protocol.

    $ stat_bolo -e tcp://localhost:2997 -l | grep 'universe'

Writing subscribers isn't that difficult, and provides an
interesting way of customizing the behavior of your monitoring
system without needing to modify bolo itself.

Next Steps
----------

Congratulations!  You have a bolo-based monitoring system up.

So where do you go from here?

You'll probably want to build a user-interface to display graphs
rendered from the RRD files, even if that's just to render the
images via cron and serve them up statically.



[ctap]:       https://github.com/filefrog/ctap
[libvigor]:   https://github.com/filefrog/libvigor
[collectors]: https://github.com/filefrog/bolo-collectors


debian/changelog
bolo (0.2.3-1) precise; urgency=low

  * Stub changelog, built by changelog-generator

 -- Package Builder <packages@niftylogic.com>  Tue, 19 May 2015 22:09:20 -0400


src/bolo.h
/*
  Copyright 2015 James Hunt <james@jameshunt.us>

  This file is part of Bolo.

  Bolo is free software: you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.

  Bolo is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along
  with Bolo.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef BOLO_H
#define BOLO_H
#include "../config.h"

#include <stdint.h>
#include <vigor.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <pthread.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <pcre.h>

#if !HAVE_PCRE_FREE_STUDY
#define pcre_free_study pcre_free
#endif

#define PACKED __attribute__((packed))

#include "bits.h"

#define OK       0
#define WARNING  1
#define CRITICAL 2
#define UNKNOWN  3
#define PENDING  4

#define DEFAULT_CONFIG_FILE "/etc/bolo.conf"
#define DEFAULT_AGENT_FILE  "/etc/dbolo.conf"

#define DEFAULT_LISTENER     "tcp://*:2999"
#define DEFAULT_CONTROLLER   "tcp://127.0.0.1:2998"
#define DEFAULT_BROADCAST    "tcp://*:2997"
#define DEFAULT_LOG_LEVEL    "error"
#define DEFAULT_LOG_FACILITY "daemon"
#define DEFAULT_RUNAS_USER   "bolo"
#define DEFAULT_RUNAS_GROUP  "bolo"
#define DEFAULT_PIDFILE      "/var/run/bolo.pid"
#define DEFAULT_SAVEFILE     "/var/lib/bolo/save.db"
#define DEFAULT_DUMPFILES    "/var/tmp/bolo.%s"

#define KERNEL_ENDPOINT "inproc://kernel"

typedef struct {
	uint16_t  freshness;
	uint8_t   status;
	char     *summary;
} type_t;

typedef struct {
	int32_t  time;

	list_t   anon;
} window_t;

typedef struct {
	type_t   *type;
	char     *name;
	int32_t   last_seen;
	int32_t   expiry;
	uint8_t   status;
	char     *summary;
	uint8_t   stale;
} state_t;

typedef struct {
	list_t      l;
	type_t     *type;

	pcre       *re;
	pcre_extra *re_extra;
} re_state_t;

typedef struct {
	window_t *window;
	char     *name;
	int32_t   last_seen;
	uint64_t  value;
} counter_t;

typedef struct {
	list_t      l;
	window_t   *window;

	pcre       *re;
	pcre_extra *re_extra;
} re_counter_t;

typedef struct {
	window_t *window;
	char     *name;
	int32_t   last_seen;

	uint64_t  n;
	double    min;
	double    max;
	double    sum;
	double    mean, mean_;
	double    var,  var_;
} sample_t;

typedef struct {
	list_t      l;
	window_t   *window;

	pcre       *re;
	pcre_extra *re_extra;
} re_sample_t;

typedef struct {
	window_t   *window;
	char       *name;
	int32_t     first_seen;
	int32_t     last_seen;

	uint64_t    first;
	uint64_t    last;
} rate_t;

typedef struct {
	list_t      l;
	window_t   *window;

	pcre       *re;
	pcre_extra *re_extra;
} re_rate_t;

typedef struct {
	list_t     l;
	int32_t    timestamp;
	char      *name;
	char      *extra;
} event_t;

typedef struct {
	hash_t  states;
	hash_t  counters;
	hash_t  samples;
	hash_t  rates;

	list_t  events;
	int     events_count;

	list_t  state_matches;
	list_t  counter_matches;
	list_t  sample_matches;
	list_t  rate_matches;

	hash_t  types;
	hash_t  windows;
	list_t  anon_windows;
} db_t;

#define EVENTS_KEEP_NUMBER 0
#define EVENTS_KEEP_TIME   1

typedef struct {
	void   *zmq;
	db_t    db;
	hash_t  keys;

	struct {
		char     *listener;
		char     *controller;
		char     *broadcast;
		uint16_t  nsca_port;

		char     *log_level;
		char     *log_facility;

		char     *pidfile;
		char     *runas_user;
		char     *runas_group;
		char     *savefile;
		char     *keysfile;
		char     *eventsfile;
		char     *dumpfiles;

		int       events_max;
		int       events_keep;
	} config;

	struct {
		uint16_t tick; /* ms */
		uint16_t freshness;
		uint16_t savestate;
	} interval;
} server_t;

#define RECORD_TYPE_MASK  0x000f
#define RECORD_TYPE_STATE    0x1
#define RECORD_TYPE_COUNTER  0x2
#define RECORD_TYPE_SAMPLE   0x3
#define RECORD_TYPE_EVENT    0x4
#define RECORD_TYPE_RATE     0x5

typedef struct PACKED {
	uint32_t  magic;
	uint16_t  version;
	uint16_t  flags;
	uint32_t  timestamp;
	uint32_t  count;
} binf_header_t;

typedef struct PACKED {
	uint16_t  len;
	uint16_t  flags;
} binf_record_t;

typedef struct PACKED {
	uint32_t  last_seen;
	 uint8_t  status;
	 uint8_t  stale;
} binf_state_t;

typedef struct PACKED {
	uint32_t  last_seen;
	uint64_t  value;
} binf_counter_t;

typedef struct PACKED {
	uint32_t  last_seen;
	uint64_t  n;
	double    min;
	double    max;
	double    sum;
	double    mean;
	double    mean_;
	double    var;
	double    var_;
} binf_sample_t;

typedef struct PACKED {
	uint32_t first_seen;
	uint32_t last_seen;
	uint64_t first;
	uint64_t last;
} binf_rate_t;

typedef struct PACKED {
	uint32_t  timestamp;
} binf_event_t;

int binf_write(db_t *db, const char *file);
int binf_read(db_t *db, const char *file);

int configure(const char *path, server_t *s);
int deconfigure(server_t *s);

void sample_reset(sample_t *sample);
int sample_data(sample_t *s, double v);

void counter_reset(counter_t *counter);

void rate_reset(rate_t *r);
int rate_data(rate_t *r, uint64_t v);
double rate_calc(rate_t *r, int32_t span);

state_t*   find_state(  db_t*, const char *name);
counter_t* find_counter(db_t*, const char *name);
sample_t*  find_sample( db_t*, const char *name);
rate_t*    find_rate(   db_t*, const char *name);

pdu_t *make_state_pdu  (int argc, char **argv, const char *ts);
pdu_t *make_counter_pdu(int argc, char **argv, const char *ts);
pdu_t *make_sample_pdu (int argc, char **argv, const char *ts);
pdu_t *make_rate_pdu   (int argc, char **argv, const char *ts);
pdu_t *make_setkeys_pdu(int argc, char **argv);
pdu_t *make_event_pdu  (int argc, char **argv, const char *ts);
pdu_t *stream_pdu(const char *line);

/* threads */
void* listener(void *u);
void* nsca_gateway(void *u);
void* controller(void *u);
void* kernel(void *u);
void* scheduler(void *u);

/* utilities - candidates for libvigor */
pdu_t *vx_pdu_dup(pdu_t *orig, const char *type);
int vx_pdu_copy(pdu_t *to, pdu_t *from, int start, int n);
int vx_timespec_parse(char *s, int modulo);
int vx_vzmq_connect(void *z, const char *endpoint);

#endif
